src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1525,148 +1525,307 @@ static LogicalResult getPartiallyFlattenedSimdCode(
 // Utilities for Op fusion at lowering
 //===----------------------------------------------------------------------===//
 
-// Function pointer type for emitScalarOpFor<T>
+// Function pointer type for the emitScalarOpFor<T> of elementwise Ops.
 typedef mlir::Value (*EmitScalarFunc)(mlir::ConversionPatternRewriter &rewriter,
     mlir::Location loc, mlir::Operation *op, mlir::Type elementType,
     mlir::ArrayRef<mlir::Value> scalarOperands);
 
-// Variadic template to iterate all the fusible Ops
+// Utility class for Op fusion.
+// Start from the root op, which is being lowered as Elementwise Op.
+// A linear sequence of fusible ops are idenitified. The operations of these ops
+// are fused into the loop nest for the root Op, and the last op is replaced
+// and the others are deleted.
+// ToFix: fusion for a graph could be done in future.
+class OpFusionHelper {
+private:
+  mlir::Operation *op_;
+  mlir::ConversionPatternRewriter &rewriter_;
+  llvm::SmallVector<mlir::Operation *, 2> fusibleOps_;
+  llvm::SmallVector<EmitScalarFunc, 2> fuseEmitFunctions_;
+
+public:
+  // Constructor
+  OpFusionHelper(
+      mlir::ConversionPatternRewriter &rewriter, mlir::Operation *startOp)
+      : op_(startOp), rewriter_(rewriter), fusibleOps_(), fuseEmitFunctions_() {
+  }
+
+  // Check whether the inputs of an op allows fusion.
+  static bool areInputsValidForFusion(Operation *producer, Operation *op);
+
+  // Check whether an Op is fusible along the use-def chain.
+  // If true, record the op and its scalar op.
+  bool checkFusibleOp(Operation *producer, Operation *op,
```

Comment:
I change all the variables to useOp and defOp  since MLIR uses define and use and it's not too long. I switched the parameter order because the useOp is usually the focus of the function, kind of the current op.